src/Common/FileSegment.cpp

```diff
@@ -0,0 +1,397 @@
+
+    if (availableSize() < size)
+        throw Exception(
+            ErrorCodes::FILE_CACHE_ERROR,
+            "Not enough space is reserved. Available: {}, expected: {}", availableSize(), size);
+
+    if (!isDownloader())
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+                        "Only downloader can do the downloading. (CallerId: {}, DownloaderId: {})",
+                        getCallerId(), downloader_id);
+
+    if (!cache_writer)
+    {
+        auto download_path = cache->path(key(), offset());
+        cache_writer = std::make_unique<WriteBufferFromFile>(download_path);
+    }
+
+    try
+    {
+        cache_writer->write(from, size);
+        cache_writer->next();
+    }
+    catch (...)
+    {
+        download_state = State::PARTIALLY_DOWNLOADED_NO_CONTINUATION;
+        throw;
+    }
+
+    downloaded_size += size;
+}
+
+FileSegment::State FileSegment::wait()
+{
+    std::unique_lock segment_lock(mutex);
+
+    if (download_state == State::EMPTY)
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Cannot wait on a file segment with empty state");
+
+    if (download_state == State::DOWNLOADING)
+    {
+        LOG_TEST(log, "{} waiting on: {}, current downloader: {}", getCallerId(), range().toString(), downloader_id);
+
+        assert(!downloader_id.empty());
+        assert(downloader_id != getCallerId());
+
+#ifndef NDEBUG
+        {
+            std::lock_guard cache_lock(cache->mutex);
+            assert(!cache->isLastFileSegmentHolder(key(), offset(), cache_lock));
+        }
+#endif
+
+        cv.wait_for(segment_lock, std::chrono::seconds(60)); /// TODO: make configurable by setting
+    }
+
+    return download_state;
+}
+
+bool FileSegment::reserve(size_t size)
+{
+    if (!size)
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Zero space reservation is not allowed");
+
+    std::lock_guard segment_lock(mutex);
+
+    if (downloaded_size + size > range().size())
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR,
+                        "Attempt to reserve space too much space ({}) for file segment with range: {} (downloaded size: {})",
+                        size, range().toString(), downloaded_size);
+
+    auto caller_id = getCallerId();
+    if (downloader_id != caller_id)
+        throw Exception(ErrorCodes::FILE_CACHE_ERROR, "Space can be reserved only by downloader (current: {}, expected: {})", caller_id, downloader_id);
+
+    assert(reserved_size >= downloaded_size);
+
+    std::lock_guard cache_lock(cache->mutex);
+
+    /**
+     * It is possible to have downloaded_size < reserved_size when reserve is called
+     * in case previous downloader did not fully download current file_segment
+     * and the caller is going to continue;
+     */
+    size_t free_space = reserved_size - downloaded_size;
+    size_t size_to_reserve = size - free_space;
+
+    bool reserved = cache->tryReserve(key(), offset(), size_to_reserve, cache_lock);
```

Comment:
We really need both `segment_lock` and `cache_lock` here?

Reply:
Yes, but updated this code a little - first lock can be released before the second.